app/uk/gov/hmrc/residencenilratebandcalculator/connectors/SessionConnector.scala

```diff
@@ -18,15 +18,117 @@ package uk.gov.hmrc.residencenilratebandcalculator.connectors
+      Constants.anyBroughtForwardAllowanceOnDisposalId -> ((v, cm) => anyBroughtForwardAllowanceOnDisposal(v, cm)))
+
+  private def clearance[A](key: String, value: A, keysToRemove: Set[String], cacheMap: CacheMap)(implicit wrts: Writes[A]): CacheMap = {
+    val mapWithDeletedKeys = cacheMap copy (data = cacheMap.data.filterKeys(s => !keysToRemove.contains(s)))
+    mapWithDeletedKeys copy (data = mapWithDeletedKeys.data + (key -> Json.toJson(value)))
+  }
+
+  private def estateHasProperty[A](value: A, cacheMap: CacheMap)(implicit wrts: Writes[A]): CacheMap =
+    clearance(Constants.estateHasPropertyId, value, Set(Constants.propertyValueId, Constants.percentageCloselyInheritedId, Constants.anyExemptionId, Constants.propertyValueAfterExemptionId), cacheMap)
+
+  private def anyExemptionClearance[A](value: A, cacheMap: CacheMap)(implicit wrts: Writes[A]): CacheMap =
+    clearance(Constants.anyExemptionId, value, Set(Constants.propertyValueAfterExemptionId), cacheMap)
+
+  private def anyBroughtForwardAllowance[A](value: A, cacheMap: CacheMap)(implicit wrts: Writes[A]): CacheMap =
+    clearance(Constants.anyBroughtForwardAllowanceId, value, Set(Constants.broughtForwardAllowanceId), cacheMap)
+
+  private def anyDownsizingAllowance[A](value: A, cacheMap: CacheMap)(implicit wrts: Writes[A]): CacheMap =
+    clearance(Constants.anyDownsizingAllowanceId,
+     value,
+     Set(
+       Constants.dateOfDisposalId,
+       Constants.valueOfDisposedPropertyId,
+       Constants.anyAssetsPassingToDirectDescendantsId,
+       Constants.assetsPassingToDirectDescendantsId,
+       Constants.anyBroughtForwardAllowanceOnDisposalId,
+       Constants.broughtForwardAllowanceOnDisposalId),
+     cacheMap)
+
+  private def anyAssetsPassingToDirectDescendants[A](value: A, cacheMap: CacheMap)(implicit wrts: Writes[A]): CacheMap =
+    clearance(Constants.anyAssetsPassingToDirectDescendantsId, value,
+      Set(
+        Constants.assetsPassingToDirectDescendantsId,
+        Constants.anyBroughtForwardAllowanceOnDisposalId,
+        Constants.broughtForwardAllowanceOnDisposalId),
+      cacheMap)
+
+  private def anyBroughtForwardAllowanceOnDisposal[A](value: A, cacheMap: CacheMap)(implicit wrts: Writes[A]): CacheMap =
+    clearance(Constants.anyBroughtForwardAllowanceOnDisposalId, value, Set(Constants.anyBroughtForwardAllowanceOnDisposalId), cacheMap)
+
+  private def updateCacheMap[A](key: String, value: A, originalCacheMap: CacheMap)(implicit wts: Writes[A]): Future[CacheMap] = {
+    val newCacheMap = funcMap.get(key).fold(originalCacheMap copy (data = originalCacheMap.data + (key -> Json.toJson(value)))) { fn => fn(Json.toJson(value), originalCacheMap)}
+    sessionRepository().upsert(newCacheMap).map {_ => newCacheMap}
+  }
+
+  def cache[A](key: String, value: A)(implicit wts: Writes[A], hc: HeaderCarrier) = {
+    hc.sessionId match {
+      case None => updateCacheMap(key, value, new CacheMap(hc.sessionId.toString, Map(key -> Json.toJson(value))))
```

Comment:
The SessionId filter should ensure we have a session Id at this point.  If we don't, I think this should be an error condition.